bigquery/storage/managedwriter/client.go

```diff
@@ -0,0 +1,222 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedwriter
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/bigquery"
+	storage "cloud.google.com/go/bigquery/storage/apiv1beta2"
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1beta2"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// StreamType indicates the type of stream this write client is managing.
+type StreamType string
+
+var (
+	// DefaultStream most closely mimics the legacy bigquery
+	// tabledata.insertAll semantics.  Successful inserts are
+	// committed immediately, and there's no tracking offsets as
+	// all writes go into a "default" stream that always exists
+	// for a table.
+	DefaultStream StreamType = "DEFAULT"
+
+	// CommittedStream appends data immediately, but creates a
+	// discrete stream for the work so that offset tracking can
+	// be used to track writes.
+	CommittedStream StreamType = "COMMITTED"
+
+	// BufferedStream is a form of checkpointed stream, that allows
+	// you to advance the offset of visible rows via Flush operations.
+	BufferedStream StreamType = "BUFFERED"
+
+	// PendingStream is a stream in which no data is made visible to
+	// readers until the stream is finalized.
+	PendingStream StreamType = "PENDING"
+)
+
+// ManagedWriteClient exposes the contract with no impl.
+type ManagedWriteClient struct {
+	streamSettings *streamSettings
+	client         *storage.BigQueryWriteClient
+}
+
+// streamSettings govern behavior of the append stream RPCs.
+type streamSettings struct {
+
+	// streamID contains the reference to the destination stream.
+	streamID string
+
+	// streamType governs behavior of the client, such as how
+	// offset handling is managed.
+	streamType StreamType
+
+	// MaxInflightRequests governs how many unacknowledged
+	// append writes can be outstanding into the system.
+	MaxInflightRequests int
+
+	// MaxInflightBytes governs how many unacknowledged
+	// request bytes can be outstanding into the system.
+	MaxInflightBytes int
+
+	// TracePrefix sets a suitable prefix for the trace ID set on
+	// append requests.  Useful for diagnostic purposes.
+	TracePrefix string
+}
+
+func defaultStreamSettings() *streamSettings {
+	return &streamSettings{
+		streamType:          DefaultStream,
+		MaxInflightRequests: 1000,
+		MaxInflightBytes:    0,
+		TracePrefix:         "defaultManagedWriter",
+	}
+}
+
+// NewManagedWriter instantiates a new managed writer.
+func NewManagedWriteClient(ctx context.Context, client *storage.BigQueryWriteClient, table *bigquery.Table, opts ...WriterOption) (*ManagedWriteClient, error) {
```

Comment:
I am not sure this should take another client as input. That means there is some pre-setup needed before calling this method. I think a function like this would be easier to use and follow the same style as other clients:

```go
func NewClient(ctx context.Context, settings SomeConfigStruct, opts ...option.ClientOption)
```

- I changed the name
- Because of dialing in this fashion we need to take clientoptions so that means writer options could not be passed as well. Used config struct.
- I left out Table. I am not sure what to do here, feedback welcome. I know for some clients we just take a string called parent and is documented and expected to be in a certain form, maybe we could do that here. It could be rolled into the settings as well I suppose. Using table make this package require the base package which I am not sure is desirable. It also has an embedded and is used in different type of semantics in the other client.

Reply:
Hmm, good feedback.  Perhaps we should have something like a NewManagedStream( opts ...WriterOption) hung off the client, which gives us multiple streams on the same client, and lets us do the client options and the writer options.  Then we move the commit off the stream and into the base client.  Let me play with this some more.

Reply:
SGTM